zrml/liquidity-mining/src/lib.rs

```diff
@@ -0,0 +1,341 @@
+            let mut initial_invalid_idx = None;
+
+            block_bought_shares.sort_unstable_by(|(.., a), (.., b)| a.cmp(&b));
+
+            for (idx, (.., bought_shares)) in block_bought_shares.iter().enumerate() {
+                let opt = total_bought_shares.checked_add(&bought_shares);
+                if let Some(el) = opt {
+                    total_bought_shares = el;
+                } else {
+                    // Adding more shares would cause an overflow
+                    initial_invalid_idx = Some(idx);
+                    break;
+                }
+            }
+
+            let range = initial_invalid_idx.unwrap_or(block_bought_shares.len())..;
+            let mut remaining_bought_shares = block_bought_shares.drain(range).collect::<Vec<_>>();
+
+            let opt = Self::buy_share_value(&per_block_distribution, &total_bought_shares);
+            let buy_share_value = if let Some(el) = opt {
+                el
+            } else {
+                // `total_bought_shares` is zero
+                return;
+            };
+
+            for (account_id, pool_id, bought_shares) in block_bought_shares {
+                let amount = if let Some(el) = buy_share_value.checked_mul(&bought_shares) {
+                    el
+                } else {
+                    // Someone is buying too many shares, beyond what `Balance` can hold. Tries again next block.
+                    remaining_bought_shares.push((account_id, pool_id, bought_shares));
+                    continue;
+                };
+
+                let rslt = <OwnedValues<T>>::try_mutate(pool_id, account_id, |values| {
+                    let err_msg = "Adding more value would cause an overflow";
+                    values.total_balance = values
+                        .total_balance
+                        .checked_add(&amount)
+                        .ok_or(DispatchError::Other(err_msg))?;
+                    values.total_shares = values
+                        .total_shares
+                        .checked_add(&bought_shares)
+                        .ok_or(DispatchError::Other(err_msg))?;
+                    Ok(())
+                });
+
+                if let Err(err) = rslt {
+                    // Account already has a lot of balance or shares. Discard incentive.
+                    Self::deposit_event(Event::BadBalanceTransfer(err));
+                }
+            }
+
+            <BlockBoughtShares<T>>::remove_all();
+            for (account, pool_id, bought_shares) in remaining_bought_shares {
+                <BlockBoughtShares<T>>::insert(account, pool_id, bought_shares);
+            }
+        }
+
+        fn remove_incentives_based_on_sold_shares() {
+            let _100_balance = BalanceOf::<T>::from(100u8);
+            let block_sold_shares = <BlockSoldShares<T>>::iter().collect::<Vec<_>>();
+
+            for (account_id, pool_id, sold_shares) in block_sold_shares {
```

Comment:
Can you explain to me please why this construct is necessary here? I mean why is 
```rust
for (account_id, pool_id, sold_shares) in <BlockSoldShares<T>>::iter() {
```
not sufficient? In the current case we iterate two times over every element if I understand it correctly.

Reply:
My bad, it doesn't. `vec` conversation only matters when calculating bought shares.